docs/design/2020-10-20-support-flushback-table-to-tso.md

```diff
@@ -0,0 +1,101 @@
+
+## Background
+
+[TiCDC](https://github.com/pingcap/ticdc) is a change data capture for TiDB, it supports to replicate change data from upstream TiDB to downstream TiDB. TiCDC supports a [tso-map feature](https://github.com/pingcap/ticdc/pull/932), which means TiCDC can replicate data to a consistent state periodically (under the conditions that upstream TiDB is available), and persists a tso map which records a consistent state between upstream and downstream. Based on that premise, when upstream meets disaster, if downstream TiDB can recover data to the latest TSO in tso-map, we could get a latest consistent state compared to upstream.
+
+We would like TiDB to provide the table flashback feature, it can be used not only in TiCDC consistent replication, but also in many data recovery scenarios, such as recovery table to eliminate recent misoperation.
+
+## Proposal
+
+### Syntax
+
+Reuse the statement syntax of existing `FLASHBACK` but add the `TIMESTAMP` keyword.
+
+```sql
+FLASHBACK TABLE tbl TIMESTAMP 420264149562425345;
+```
+
+In the feture we could probably extend the syntax to support flashback a database or all databases.
+
+```sql
+FLASHBACK DATABASE db TIMESTAMP 420264149562425345;
+FLASHBACK DATABASE * TIMESTAMP 420264149562425345;
+```
+
+### Execution
+
+The `FLASHBACK x TIMESTAMP` statement is **synchronous**, it blocks the current session until completion. Nevertheless, the query can be canceled using the standard `KILL TIDB QUERY n` statement.
+
+The `FLASHBACK x TIMESTAMP` statement interacts with TiKV to read data, and could write to TiKV via SQL execute API or KV API, so the statement requires SUPER privilege to execute.
+
+### Status query
+
+The status of the queued and running tasks can be queried using the `SHOW` statement:
+
+```
+SHOW FLASHBACK;
+```
+
+Example:
+
+```
+MySQL [(none)]> show backups;
++------------+---------+----------+---------------------+---------------------+---------------------+
+| Table      | State   | Progress | Queue_Time          | Execution_Time      | TSO                 |
++------------+---------+----------+---------------------+---------------------+---------------------+
+| test_db.t1 | Scan    | 20.01    | 2020-10-12 23:09:03 | 2020-10-12 23:09:25 | 420264149562425345  |
++------------+---------+----------+---------------------+---------------------+---------------------+
+```
+
+## Rationale
+
+Currently the `FLASHBACK x TIMESTAMP` only supports TiKV as backend store. TiKV provides a gRPC API [EventFeed](https://github.com/pingcap/kvproto/blob/cdcb788eaebd513df2e5e98a7b8b4fe6132713cc/proto/cdcpb.proto#L146) which can return all raw KV changed log of a region from given TSO to now. Besides each KV changed log can provide both the before row value and after row value if the `old-value` switch is turned on. With the help of old value feature, it is easily to distinguish the DML type including INSERT/UPDATE/DELETE, and revert data from new value to old value.
```

Comment:
What is the expected performance of `FLASHBACK`? Looks it may be slow because:

- We need to retrieve the changelogs and convert them to the 'undo SQLs'.
- (Probably?) The SQLs are required to run in serial.

Reply:
undo SQLs can be executed concurrently if no pk and unique key conflicts.
Besides we could omit one step from datum to undo SQL, and use datum to write data directly.

Reply:
How 'pk and unique key conflict' may happen? From concurrent DMLs, if there is, right?

Reply:
yes, a sequence of DMLs can be executed concurrently if there is no causality conflict. refer to https://github.com/pingcap/ticdc/issues/463